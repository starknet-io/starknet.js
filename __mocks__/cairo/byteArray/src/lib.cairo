// ByteArray Storage Contract Interface
#[starknet::interface]
pub trait IByteArrayStorage<TContractState> {
    fn store_message(ref self: TContractState, message: ByteArray);
    fn read_message(self: @TContractState) -> ByteArray;
}

// ByteArray Storage Contract
#[starknet::contract]
pub mod ByteArrayStorage {
    use starknet::get_caller_address;
    use starknet::storage::*;

    #[storage]
    struct Storage {
        stored_message: ByteArray,
    }

    #[event]
    #[derive(Drop, starknet::Event)]
    pub enum Event {
        MessageStored: MessageStored,
    }

    #[derive(Drop, starknet::Event)]
    pub struct MessageStored {
        pub caller: starknet::ContractAddress,
        pub message: ByteArray,
    }

    #[constructor]
    fn constructor(ref self: ContractState) {
        // Initialize with empty ByteArray
        self.stored_message.write("");
    }

    #[abi(embed_v0)]
    impl ByteArrayStorageImpl of super::IByteArrayStorage<ContractState> {
        fn store_message(ref self: ContractState, message: ByteArray) {
            let caller = get_caller_address();
            
            // Store the message in storage
            self.stored_message.write(message.clone());
            
            // Emit event with the message
            self.emit(Event::MessageStored(MessageStored { 
                caller, 
                message 
            }));
        }

        fn read_message(self: @ContractState) -> ByteArray {
            self.stored_message.read()
        }
    }
}