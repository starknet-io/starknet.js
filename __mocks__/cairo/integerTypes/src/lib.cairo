#[starknet::interface]
trait IIntegerTypesStorage<TContractState> {
    // Unsigned integer functions
    fn store_u8(ref self: TContractState, value: u8);
    fn read_u8(self: @TContractState) -> u8;
    fn store_u16(ref self: TContractState, value: u16);
    fn read_u16(self: @TContractState) -> u16;
    fn store_u64(ref self: TContractState, value: u64);
    fn read_u64(self: @TContractState) -> u64;
    fn store_u128(ref self: TContractState, value: u128);
    fn read_u128(self: @TContractState) -> u128;

    // Signed integer functions
    fn store_i8(ref self: TContractState, value: i8);
    fn read_i8(self: @TContractState) -> i8;
    fn store_i16(ref self: TContractState, value: i16);
    fn read_i16(self: @TContractState) -> i16;
    fn store_i32(ref self: TContractState, value: i32);
    fn read_i32(self: @TContractState) -> i32;
    fn store_i64(ref self: TContractState, value: i64);
    fn read_i64(self: @TContractState) -> i64;
    fn store_i128(ref self: TContractState, value: i128);
    fn read_i128(self: @TContractState) -> i128;

    // Batch operations
    fn store_all_unsigned(
        ref self: TContractState,
        u8_val: u8,
        u16_val: u16,
        u64_val: u64,
        u128_val: u128
    );
    fn read_all_unsigned(self: @TContractState) -> (u8, u16, u64, u128);
    fn store_all_signed(
        ref self: TContractState,
        i8_val: i8,
        i16_val: i16,
        i32_val: i32,
        i64_val: i64,
        i128_val: i128
    );
    fn read_all_signed(self: @TContractState) -> (i8, i16, i32, i64, i128);

    // Boundary value testing
    fn test_boundary_values_unsigned(self: @TContractState) -> (u8, u16, u64, u128);
    fn test_boundary_values_signed(self: @TContractState) -> (i8, i16, i32, i64, i128);
    fn test_negative_boundary_values_signed(self: @TContractState) -> (i8, i16, i32, i64, i128);
}

#[starknet::contract]
mod IntegerTypesStorage {
    use starknet::storage::{StoragePointerReadAccess, StoragePointerWriteAccess};

    #[storage]
    struct Storage {
        // Unsigned integer types
        stored_u8: u8,
        stored_u16: u16,
        stored_u64: u64,
        stored_u128: u128,
        // Signed integer types
        stored_i8: i8,
        stored_i16: i16,
        stored_i32: i32,
        stored_i64: i64,
        stored_i128: i128,
    }

    #[event]
    #[derive(Drop, starknet::Event)]
    enum Event {
        // Unsigned integer events
        U8Stored: U8Stored,
        U16Stored: U16Stored,
        U64Stored: U64Stored,
        U128Stored: U128Stored,
        // Signed integer events
        I8Stored: I8Stored,
        I16Stored: I16Stored,
        I32Stored: I32Stored,
        I64Stored: I64Stored,
        I128Stored: I128Stored,
    }

    #[derive(Drop, starknet::Event)]
    struct U8Stored {
        #[key]
        value: u8,
    }

    #[derive(Drop, starknet::Event)]
    struct U16Stored {
        #[key]
        value: u16,
    }

    #[derive(Drop, starknet::Event)]
    struct U64Stored {
        #[key]
        value: u64,
    }

    #[derive(Drop, starknet::Event)]
    struct U128Stored {
        #[key]
        value: u128,
    }

    #[derive(Drop, starknet::Event)]
    struct I8Stored {
        #[key]
        value: i8,
    }

    #[derive(Drop, starknet::Event)]
    struct I16Stored {
        #[key]
        value: i16,
    }

    #[derive(Drop, starknet::Event)]
    struct I32Stored {
        #[key]
        value: i32,
    }

    #[derive(Drop, starknet::Event)]
    struct I64Stored {
        #[key]
        value: i64,
    }

    #[derive(Drop, starknet::Event)]
    struct I128Stored {
        #[key]
        value: i128,
    }

    #[abi(embed_v0)]
    impl IntegerTypesStorageImpl of super::IIntegerTypesStorage<ContractState> {
        // Unsigned integer storage functions
        fn store_u8(ref self: ContractState, value: u8) {
            self.stored_u8.write(value);
            self.emit(U8Stored { value });
        }

        fn read_u8(self: @ContractState) -> u8 {
            self.stored_u8.read()
        }

        fn store_u16(ref self: ContractState, value: u16) {
            self.stored_u16.write(value);
            self.emit(U16Stored { value });
        }

        fn read_u16(self: @ContractState) -> u16 {
            self.stored_u16.read()
        }

        fn store_u64(ref self: ContractState, value: u64) {
            self.stored_u64.write(value);
            self.emit(U64Stored { value });
        }

        fn read_u64(self: @ContractState) -> u64 {
            self.stored_u64.read()
        }


        fn store_u128(ref self: ContractState, value: u128) {
            self.stored_u128.write(value);
            self.emit(U128Stored { value });
        }

        fn read_u128(self: @ContractState) -> u128 {
            self.stored_u128.read()
        }

        // Signed integer storage functions
        fn store_i8(ref self: ContractState, value: i8) {
            self.stored_i8.write(value);
            self.emit(I8Stored { value });
        }

        fn read_i8(self: @ContractState) -> i8 {
            self.stored_i8.read()
        }

        fn store_i16(ref self: ContractState, value: i16) {
            self.stored_i16.write(value);
            self.emit(I16Stored { value });
        }

        fn read_i16(self: @ContractState) -> i16 {
            self.stored_i16.read()
        }

        fn store_i32(ref self: ContractState, value: i32) {
            self.stored_i32.write(value);
            self.emit(I32Stored { value });
        }

        fn read_i32(self: @ContractState) -> i32 {
            self.stored_i32.read()
        }

        fn store_i64(ref self: ContractState, value: i64) {
            self.stored_i64.write(value);
            self.emit(I64Stored { value });
        }

        fn read_i64(self: @ContractState) -> i64 {
            self.stored_i64.read()
        }

        fn store_i128(ref self: ContractState, value: i128) {
            self.stored_i128.write(value);
            self.emit(I128Stored { value });
        }

        fn read_i128(self: @ContractState) -> i128 {
            self.stored_i128.read()
        }

        // Batch operations for testing multiple types at once
        fn store_all_unsigned(
            ref self: ContractState,
            u8_val: u8,
            u16_val: u16,
            u64_val: u64,
            u128_val: u128
        ) {
            self.store_u8(u8_val);
            self.store_u16(u16_val);
            self.store_u64(u64_val);
            self.store_u128(u128_val);
        }

        fn read_all_unsigned(self: @ContractState) -> (u8, u16, u64, u128) {
            (
                self.read_u8(),
                self.read_u16(),
                self.read_u64(),
                self.read_u128()
            )
        }

        fn store_all_signed(
            ref self: ContractState,
            i8_val: i8,
            i16_val: i16,
            i32_val: i32,
            i64_val: i64,
            i128_val: i128
        ) {
            self.store_i8(i8_val);
            self.store_i16(i16_val);
            self.store_i32(i32_val);
            self.store_i64(i64_val);
            self.store_i128(i128_val);
        }

        fn read_all_signed(self: @ContractState) -> (i8, i16, i32, i64, i128) {
            (
                self.read_i8(),
                self.read_i16(),
                self.read_i32(),
                self.read_i64(),
                self.read_i128()
            )
        }

        // Test boundary values
        fn test_boundary_values_unsigned(self: @ContractState) -> (u8, u16, u64, u128) {
            (
                255_u8,            // Max u8
                65535_u16,         // Max u16
                18446744073709551615_u64, // Max u64
                340282366920938463463374607431768211455_u128 // Max u128
            )
        }

        fn test_boundary_values_signed(self: @ContractState) -> (i8, i16, i32, i64, i128) {
            (
                127_i8,            // Max i8
                32767_i16,         // Max i16
                2147483647_i32,    // Max i32
                9223372036854775807_i64, // Max i64
                170141183460469231731687303715884105727_i128 // Max i128
            )
        }

        fn test_negative_boundary_values_signed(self: @ContractState) -> (i8, i16, i32, i64, i128) {
            (
                -128_i8,           // Min i8
                -32768_i16,        // Min i16
                -2147483648_i32,   // Min i32
                -9223372036854775808_i64, // Min i64
                -170141183460469231731687303715884105728_i128 // Min i128
            )
        }
    }
}