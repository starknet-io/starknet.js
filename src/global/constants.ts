import type { FeeMarginPercentage } from '../types';
import { ETransactionVersion, RPCSPEC08 } from '../types/api';
import type { LogLevel } from './logger.type';

export { IS_BROWSER } from '../utils/encode';

/**
 * Cairo Felt support storing max 31 character
 */
export const TEXT_TO_FELT_MAX_LEN = 31;

/**
 * Alternatively use directly from api specification
 * types.RPC.ETransactionVersion
 * For BN do BigInt(TRANSACTION_VERSION.*)
 */
export const { ETransactionVersion: TRANSACTION_VERSION } = RPCSPEC08;

export const ZERO = 0n;
export const MASK_250 = 2n ** 250n - 1n; // 2 ** 250 - 1
export const MASK_31 = 2n ** 31n - 1n; // 2 ** 31 - 1
export const API_VERSION = ZERO;
export const PRIME = 2n ** 251n + 17n * 2n ** 192n + 1n;

// based on: https://github.com/starkware-libs/cairo-lang/blob/v0.12.3/src/starkware/starknet/common/storage.cairo#L3
export const MAX_STORAGE_ITEM_SIZE = 256n;
export const ADDR_BOUND = 2n ** 251n - MAX_STORAGE_ITEM_SIZE;

const range = (min: bigint, max: bigint) => ({ min, max }) as const;

export const RANGE_FELT = range(ZERO, PRIME - 1n);
export const RANGE_I128 = range(-(2n ** 127n), 2n ** 127n - 1n);
export const RANGE_U128 = range(ZERO, 2n ** 128n - 1n);

export enum BaseUrl {
  SN_MAIN = 'https://alpha-mainnet.starknet.io',
  SN_SEPOLIA = 'https://alpha-sepolia.starknet.io',
}

export enum NetworkName {
  SN_MAIN = 'SN_MAIN',
  SN_SEPOLIA = 'SN_SEPOLIA',
}

export enum StarknetChainId {
  SN_MAIN = '0x534e5f4d41494e', // encodeShortString('SN_MAIN'),
  SN_SEPOLIA = '0x534e5f5345504f4c4941', // encodeShortString('SN_SEPOLIA')
}

export enum TransactionHashPrefix {
  DECLARE = '0x6465636c617265', // encodeShortString('declare'),
  DEPLOY = '0x6465706c6f79', // encodeShortString('deploy'),
  DEPLOY_ACCOUNT = '0x6465706c6f795f6163636f756e74', // encodeShortString('deploy_account'),
  INVOKE = '0x696e766f6b65', // encodeShortString('invoke'),
  L1_HANDLER = '0x6c315f68616e646c6572', // encodeShortString('l1_handler'),
}

export const UDC = {
  ADDRESS: '0x041a78e741e5af2fec34b695679bc6891742439f7afb8484ecd7766661ad02bf',
  ENTRYPOINT: 'deployContract',
} as const;

export const OutsideExecutionCallerAny = '0x414e595f43414c4c4552'; // encodeShortString('ANY_CALLER')
export const SNIP9_V1_INTERFACE_ID =
  '0x68cfd18b92d1907b8ba3cc324900277f5a3622099431ea85dd8089255e4181';
export const SNIP9_V2_INTERFACE_ID =
  '0x1d1144bb2138366ff28d8e9ab57456b1d332ac42196230c3a602003c89872';

// Ledger signer
// 0x80
export const HARDENING_BYTE = 128;
// 0x80000000
export const HARDENING_4BYTES = 2147483648n;

/**
 * dot formate rpc versions
 */
export const SupportedRpcVersion = {
  0.7: '0.7',
  0.8: '0.8',
} as const;
export type SupportedRpcVersion = (typeof SupportedRpcVersion)[keyof typeof SupportedRpcVersion];

export type SupportedTransactionVersion =
  | typeof ETransactionVersion.V2
  | typeof ETransactionVersion.V3;

// Default initial global config
export const DEFAULT_GLOBAL_CONFIG: {
  legacyMode: boolean;
  logLevel: LogLevel;
  rpcVersion: SupportedRpcVersion;
  transactionVersion: SupportedTransactionVersion;
  feeMarginPercentage: FeeMarginPercentage;
} = {
  legacyMode: false,
  rpcVersion: '0.8',
  transactionVersion: ETransactionVersion.V3,
  logLevel: 'INFO',
  feeMarginPercentage: {
    bounds: {
      l1_gas: {
        max_amount: 50,
        max_price_per_unit: 50,
      },
      l1_data_gas: {
        max_amount: 50,
        max_price_per_unit: 50,
      },
      l2_gas: {
        max_amount: 50,
        max_price_per_unit: 50,
      },
    },
    maxFee: 50,
  },
};

const vToUrl = (versionString: SupportedRpcVersion) =>
  `v${versionString.replace(/^v/, '').replace(/\./g, '_')}`;

export const RPC_NODES = {
  SN_MAIN: [
    `https://starknet-mainnet.public.blastapi.io/rpc/${vToUrl(DEFAULT_GLOBAL_CONFIG.rpcVersion)}`,
    `https://free-rpc.nethermind.io/mainnet-juno/${vToUrl(DEFAULT_GLOBAL_CONFIG.rpcVersion)}`,
  ],
  SN_SEPOLIA: [
    `https://starknet-sepolia.public.blastapi.io/rpc/${vToUrl(DEFAULT_GLOBAL_CONFIG.rpcVersion)}`,
    `https://free-rpc.nethermind.io/sepolia-juno/${vToUrl(DEFAULT_GLOBAL_CONFIG.rpcVersion)}`,
  ],
} as const;

// Default system messages
export const SYSTEM_MESSAGES = {
  legacyTxWarningMessage:
    'You are using a deprecated transaction version (V0,V1,V2)!\nUpdate to the latest V3 transactions!',
  legacyTxRPC08Message: 'RPC 0.8 do not support legacy transactions',
  SWOldV3: 'RPC 0.7 V3 tx (improper resource bounds) not supported in RPC 0.8',
};
