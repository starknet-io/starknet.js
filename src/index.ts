/**
 * Main
 */
export * from './wallet';
export * from './account';
export * from './contract';
export * from './provider';
export * from './signer';
export * from './channel';

// TODO: decide on final export style
export * from './types';
export * as types from './types';

/**
 * Utils
 */
export * as constants from './global/constants';
export * as encode from './utils/encode';
export * as hash from './utils/hash';
export * as v3hash from './utils/hash/transactionHash/v3';
export * as v2hash from './utils/hash/transactionHash/v2';
export * as json from './utils/json';
export * as num from './utils/num';
export * as transaction from './utils/transaction';
export * as stark from './utils/stark';
export * as eth from './utils/eth';
export * as merkle from './utils/merkle';
export * as uint256 from './utils/uint256';
export * as shortString from './utils/shortString';
export * as typedData from './utils/typedData';
export * as ec from './utils/ec';
export * as starknetId from './utils/starknetId';
export * as provider from './utils/provider';
export * as selector from './utils/hash/selector';
export * as events from './utils/events';
export * as outsideExecution from './utils/outsideExecution';
export * as src5 from './utils/src5';
export * from './utils/batch';
export * from './utils/responseParser';
export * from './utils/cairoDataTypes/uint256';
export * from './utils/cairoDataTypes/uint512';
export * from './utils/cairoDataTypes/fixedArray';
export * from './utils/address';
export * from './utils/calldata';
export * from './utils/calldata/enum';
export * from './utils/contract';
export * from './utils/transactionReceipt';
export * from './utils/units';
export * as wallet from './wallet/connect';
export * from './global/config';
export * from './global/logger';
export * from './global/logger.type';

/**
 * Deprecated
 */
/* eslint-disable import/first */
import * as num from './utils/num';

/** @deprecated prefer the 'num' naming */
export const number = num;
