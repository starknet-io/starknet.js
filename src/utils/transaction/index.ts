export * from './getCompiledCalldata';
export * from './transaction';
