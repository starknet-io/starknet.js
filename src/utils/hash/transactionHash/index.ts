/**
 * Transaction Hash based on Transaction Version
 */

import { StarknetChainId } from '../../../global/constants';
import { ResourceBounds } from '../../../provider/types/spec.type';
import { BigNumberish, Calldata } from '../../../types';
import {
  EDAMode,
  ETransactionVersion,
  ETransactionVersion2,
  ETransactionVersion3,
} from '../../../types/api';
import {
  calculateDeclareTransactionHash as v2calculateDeclareTransactionHash,
  calculateDeployAccountTransactionHash as v2calculateDeployAccountTransactionHash,
  calculateTransactionHash as v2calculateInvokeTransactionHash,
} from './v2';
import {
  calculateDeclareTransactionHash as v3calculateDeclareTransactionHash,
  calculateDeployAccountTransactionHash as v3calculateDeployAccountTransactionHash,
  calculateInvokeTransactionHash as v3calculateInvokeTransactionHash,
} from './v3';

export { calculateL2MessageTxHash } from './v2';
/*
 * INVOKE TX HASH
 */

type Version = typeof ETransactionVersion.V3 | typeof ETransactionVersion.F3;

function isV3InvokeTx(args: CalcInvokeTxHashArgs): args is CalcV3InvokeTxHashArgs {
  return [ETransactionVersion.V3, ETransactionVersion.F3].includes(args.version as Version);
}

type CalcV2InvokeTxHashArgs = {
  senderAddress: BigNumberish;
  version: `${ETransactionVersion2}`;
  compiledCalldata: Calldata;
  maxFee: BigNumberish;
  chainId: StarknetChainId;
  nonce: BigNumberish;
};

type CalcV3InvokeTxHashArgs = {
  senderAddress: BigNumberish;
  version: `${ETransactionVersion3}`;
  compiledCalldata: Calldata;
  chainId: StarknetChainId;
  nonce: BigNumberish;
  accountDeploymentData: BigNumberish[];
  nonceDataAvailabilityMode: EDAMode;
  feeDataAvailabilityMode: EDAMode;
  resourceBounds: ResourceBounds;
  tip: BigNumberish;
  paymasterData: BigNumberish[];
};

type CalcInvokeTxHashArgs = CalcV2InvokeTxHashArgs | CalcV3InvokeTxHashArgs;

export function calculateInvokeTransactionHash(args: CalcInvokeTxHashArgs) {
  if (isV3InvokeTx(args)) {
    return v3calculateInvokeTransactionHash(
      args.senderAddress,
      args.version,
      args.compiledCalldata,
      args.chainId,
      args.nonce,
      args.accountDeploymentData,
      args.nonceDataAvailabilityMode,
      args.feeDataAvailabilityMode,
      args.resourceBounds,
      args.tip,
      args.paymasterData
    );
  }
  return v2calculateInvokeTransactionHash(
    args.senderAddress,
    args.version,
    args.compiledCalldata,
    args.maxFee,
    args.chainId,
    args.nonce
  );
}

/*
 * DECLARE TX HASH
 */
function isV3DeclareTx(args: CalcDeclareTxHashArgs): args is CalcV3DeclareTxHashArgs {
  return [ETransactionVersion.V3, ETransactionVersion.F3].includes(args.version as Version);
}

type CalcV2DeclareTxHashArgs = {
  classHash: string;
  senderAddress: BigNumberish;
  version: `${ETransactionVersion2}`;
  maxFee: BigNumberish;
  chainId: StarknetChainId;
  nonce: BigNumberish;
  compiledClassHash?: string;
};

type CalcV3DeclareTxHashArgs = {
  classHash: string;
  compiledClassHash: string;
  senderAddress: BigNumberish;
  version: `${ETransactionVersion3}`;
  chainId: StarknetChainId;
  nonce: BigNumberish;
  accountDeploymentData: BigNumberish[];
  nonceDataAvailabilityMode: EDAMode;
  feeDataAvailabilityMode: EDAMode;
  resourceBounds: ResourceBounds;
  tip: BigNumberish;
  paymasterData: BigNumberish[];
};

type CalcDeclareTxHashArgs = CalcV2DeclareTxHashArgs | CalcV3DeclareTxHashArgs;

export function calculateDeclareTransactionHash(args: CalcDeclareTxHashArgs) {
  if (isV3DeclareTx(args)) {
    return v3calculateDeclareTransactionHash(
      args.classHash,
      args.compiledClassHash,
      args.senderAddress,
      args.version,
      args.chainId,
      args.nonce,
      args.accountDeploymentData,
      args.nonceDataAvailabilityMode,
      args.feeDataAvailabilityMode,
      args.resourceBounds,
      args.tip,
      args.paymasterData
    );
  }

  return v2calculateDeclareTransactionHash(
    args.classHash,
    args.senderAddress,
    args.version,
    args.maxFee,
    args.chainId,
    args.nonce,
    args.compiledClassHash
  );
}

/*
 * DEPLOY ACCOUNT TX HASH
 */

function isV3DeployAccountTx(
  args: CalcDeployAccountTxHashArgs
): args is CalcV3DeployAccountTxHashArgs {
  return [ETransactionVersion.V3, ETransactionVersion.F3].includes(args.version as Version);
}

type CalcV2DeployAccountTxHashArgs = {
  contractAddress: BigNumberish;
  classHash: BigNumberish;
  constructorCalldata: Calldata;
  salt: BigNumberish;
  version: `${ETransactionVersion2}`;
  maxFee: BigNumberish;
  chainId: StarknetChainId;
  nonce: BigNumberish;
};

type CalcV3DeployAccountTxHashArgs = {
  contractAddress: BigNumberish;
  classHash: BigNumberish;
  compiledConstructorCalldata: Calldata;
  salt: BigNumberish;
  version: `${ETransactionVersion3}`;
  chainId: StarknetChainId;
  nonce: BigNumberish;
  nonceDataAvailabilityMode: EDAMode;
  feeDataAvailabilityMode: EDAMode;
  resourceBounds: ResourceBounds;
  tip: BigNumberish;
  paymasterData: BigNumberish[];
};

type CalcDeployAccountTxHashArgs = CalcV2DeployAccountTxHashArgs | CalcV3DeployAccountTxHashArgs;

export function calculateDeployAccountTransactionHash(args: CalcDeployAccountTxHashArgs) {
  if (isV3DeployAccountTx(args)) {
    return v3calculateDeployAccountTransactionHash(
      args.contractAddress,
      args.classHash,
      args.compiledConstructorCalldata,
      args.salt,
      args.version,
      args.chainId,
      args.nonce,
      args.nonceDataAvailabilityMode,
      args.feeDataAvailabilityMode,
      args.resourceBounds,
      args.tip,
      args.paymasterData
    );
  }

  return v2calculateDeployAccountTransactionHash(
    args.contractAddress,
    args.classHash,
    args.constructorCalldata,
    args.salt,
    args.version,
    args.maxFee,
    args.chainId,
    args.nonce
  );
}
