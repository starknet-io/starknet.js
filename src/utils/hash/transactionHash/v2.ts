/**
 * Calculate Hashes for v0 - v2 transactions
 */

/* eslint-disable no-param-reassign */
/* eslint-disable import/extensions */
import { StarknetChainId, TransactionHashPrefix } from '../../../global/constants';
import { BigNumberish, RawCalldata } from '../../../types';
import { starkCurve } from '../../ec';
import { toBigInt } from '../../num';
import { getSelector } from '../selector';

/**
 * Compute pedersen hash from data
 * @returns format: hex-string - pedersen hash
 */
export function computeHashOnElements(data: BigNumberish[]): string {
  return [...data, data.length]
    .reduce((x: BigNumberish, y: BigNumberish) => starkCurve.pedersen(toBigInt(x), toBigInt(y)), 0)
    .toString();
}

/**
 * Calculate transaction pedersen hash for common properties
 *
 * Following implementation is based on this python [implementation #](https://github.com/starkware-libs/cairo-lang/blob/b614d1867c64f3fb2cf4a4879348cfcf87c3a5a7/src/starkware/starknet/core/os/transaction_hash/transaction_hash.py)
 * @returns format: hex-string
 */
export function calculateTransactionHashCommon(
  txHashPrefix: TransactionHashPrefix,
  version: BigNumberish,
  contractAddress: BigNumberish,
  entryPointSelector: BigNumberish,
  calldata: RawCalldata,
  maxFee: BigNumberish,
  chainId: StarknetChainId,
  additionalData: BigNumberish[] = []
): string {
  const calldataHash = computeHashOnElements(calldata);
  const dataToHash = [
    txHashPrefix,
    version,
    contractAddress,
    entryPointSelector,
    calldataHash,
    maxFee,
    chainId,
    ...additionalData,
  ];
  return computeHashOnElements(dataToHash);
}

/**
 * Calculate declare transaction hash
 * @param classHash hex-string
 * @param compiledClassHash hex-string
 * @returns format: hex-string
 */
export function calculateDeclareTransactionHash(
  classHash: string,
  senderAddress: BigNumberish,
  version: BigNumberish,
  maxFee: BigNumberish,
  chainId: StarknetChainId,
  nonce: BigNumberish,
  compiledClassHash?: string
): string {
  return calculateTransactionHashCommon(
    TransactionHashPrefix.DECLARE,
    version,
    senderAddress,
    0,
    [classHash],
    maxFee,
    chainId,
    [nonce, ...(compiledClassHash ? [compiledClassHash] : [])]
  );
}

/**
 * Calculate deploy_account transaction hash
 * @returns format: hex-string
 */
export function calculateDeployAccountTransactionHash(
  contractAddress: BigNumberish,
  classHash: BigNumberish,
  constructorCalldata: RawCalldata,
  salt: BigNumberish,
  version: BigNumberish,
  maxFee: BigNumberish,
  chainId: StarknetChainId,
  nonce: BigNumberish
) {
  const calldata = [classHash, salt, ...constructorCalldata];

  return calculateTransactionHashCommon(
    TransactionHashPrefix.DEPLOY_ACCOUNT,
    version,
    contractAddress,
    0,
    calldata,
    maxFee,
    chainId,
    [nonce]
  );
}

/**
 * Calculate invoke transaction hash
 * @returns format: hex-string
 */
export function calculateTransactionHash(
  contractAddress: BigNumberish,
  version: BigNumberish,
  calldata: RawCalldata,
  maxFee: BigNumberish,
  chainId: StarknetChainId,
  nonce: BigNumberish
): string {
  return calculateTransactionHashCommon(
    TransactionHashPrefix.INVOKE,
    version,
    contractAddress,
    0,
    calldata,
    maxFee,
    chainId,
    [nonce]
  );
}

/**
 * Calculate the L2 transaction hash generated by a message L1->L2
 * @param {BigNumberish} l1FromAddress L1 account address that paid the message.
 * @param {BigNumberish} l2ToAddress L2 contract address to execute.
 * @param {string | BigNumberish} l2Selector can be a function name ("bridge_withdraw") or a number (BigNumberish).
 * @param {RawCalldata} l2Calldata an array of BigNumberish of the raw parameters passed to the above function.
 * @param {BigNumberish} l2ChainId L2 chain ID : from constants.StarknetChainId.xxx
 * @param {BigNumberish} l1Nonce The nonce of the L1 account.
 * @returns {string} hex-string of the L2 transaction hash
 * @example
 * ```typescript
 * const l1FromAddress = "0x0000000000000000000000008453fc6cd1bcfe8d4dfc069c400b433054d47bdc";
 * const l2ToAddress = 2158142789748719025684046545159279785659305214176670733242887773692203401023n;
 * const l2Selector = 774397379524139446221206168840917193112228400237242521560346153613428128537n;
 * const payload = [
 *     4543560n,
 *    829565602143178078434185452406102222830667255948n,
 *     3461886633118033953192540141609307739580461579986333346825796013261542798665n,
 *     9000000000000000n,
 *     0n,
 * ];
 * const l1Nonce = 8288n;
 * const result = hash.calculateL2MessageTxHash(l1FromAddress, l2ToAddress, l2Selector, payload, constants.StarknetChainId.SN_SEPOLIA, l1Nonce);
 * // result = "0x67d959200d65d4ad293aa4b0da21bb050a1f669bce37d215c6edbf041269c07"
 * ```
 */
export function calculateL2MessageTxHash(
  l1FromAddress: BigNumberish,
  l2ToAddress: BigNumberish,
  l2Selector: string | BigNumberish,
  l2Calldata: RawCalldata,
  l2ChainId: StarknetChainId,
  l1Nonce: BigNumberish
): string {
  const payload = [l1FromAddress, ...l2Calldata];
  return calculateTransactionHashCommon(
    TransactionHashPrefix.L1_HANDLER,
    0,
    l2ToAddress,
    getSelector(l2Selector),
    payload,
    0,
    l2ChainId,
    [l1Nonce]
  );
}
