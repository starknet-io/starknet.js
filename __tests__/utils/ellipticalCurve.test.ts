import { constants, ec } from '../../src';
import { StarknetChainId } from '../../src/global/constants';
import { computeHashOnElements } from '../../src/utils/hash';
import { calculateTransactionHash } from '../../src/utils/hash/transactionHash/v2';
import { fromCallsToExecuteCalldata } from '../../src/utils/transaction';

test('getKeyPair()', () => {
  const privateKey = '0x019800ea6a9a73f94aee6a3d2edf018fc770443e90c7ba121e8303ec6b349279';
  const starkKey = ec.starkCurve.getStarkKey(privateKey);
  // somehow needed, returns error else
  expect(starkKey).toBe('0x33f45f07e1bd1a51b45fc24ec8c8c9908db9e42191be9e169bfcac0c0d99745');
});

test('pedersen()', () => {
  const own = ec.starkCurve.pedersen('0x12773', '0x872362');
  expect(own).toMatchSnapshot();
});

test('pedersen() with 0', () => {
  const own = ec.starkCurve.pedersen('0x12773', '0x0');
  expect(own).toMatchSnapshot();
});

test('computeHashOnElements()', () => {
  const array = ['1', '2', '3', '4'];
  expect(computeHashOnElements(array)).toBe(
    '0x66bd4335902683054d08a0572747ea78ebd9e531536fb43125424ca9f902084'
  );
  expect(array).toStrictEqual(['1', '2', '3', '4']);

  expect(computeHashOnElements(['1', '2'])).toBe(
    '0x501a3a8e6cd4f5241c639c74052aaa34557aafa84dd4ba983d6443c590ab7df'
  );
});

test('hashMessage()', () => {
  const privateKey = '0x019800ea6a9a73f94aee6a3d2edf018fc770443e90c7ba121e8303ec6b349279';
  const account = '2007067565103695475819120104515800035851923905855118399071773059478896040938';
  const transactions = [
    {
      contractAddress:
        '3290661298119599979891444342541795905081168856323302956721669397616389152866',
      entrypoint: 'set_number',
      calldata: ['47'],
    },
  ];
  const nonce = '3';
  const maxFee = '0';
  const calldata = [...fromCallsToExecuteCalldata(transactions), nonce];

  const hashMsg = calculateTransactionHash(
    account,
    BigInt(constants.TRANSACTION_VERSION.V1),
    calldata,
    maxFee,
    StarknetChainId.SN_SEPOLIA,
    nonce
  );

  expect(hashMsg).toMatchInlineSnapshot(
    `"0xa006ce6da518722c1af8bdb1d8a42cee638102c670bb1a55f063bff10506d4"`
  );

  const { r, s } = ec.starkCurve.sign(hashMsg, privateKey);

  expect(r.toString()).toMatchInlineSnapshot(
    `"384207128292005766686294801921397180350977625816434242436096267488258549139"`
  );
  expect(s.toString()).toMatchInlineSnapshot(
    `"2521602681140573534692734854765316415611209530542226558354401890884906162365"`
  );
});

test('verify signed message()', () => {
  const pk = '0x019800ea6a9a73f94aee6a3d2edf018fc770443e90c7ba121e8303ec6b349279';
  const account = '0x33f45f07e1bd1a51b45fc24ec8c8c9908db9e42191be9e169bfcac0c0d99745';
  const price = '1';
  const hashMsg = ec.starkCurve.pedersen(account, price);
  const signature = ec.starkCurve.sign(hashMsg, pk);
  const pubKey = ec.starkCurve.getPublicKey(pk);
  expect(ec.starkCurve.verify(signature.toDERHex(), hashMsg, pubKey)).toBe(true);
});
