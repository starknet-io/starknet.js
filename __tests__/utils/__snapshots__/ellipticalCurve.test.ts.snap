// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`pedersen() 1`] = `"0x5ed2703dfdb505c587700ce2ebfcab5b3515cd7e6114817e6026ec9d4b364ca"`;

exports[`pedersen() with 0 1`] = `"0x1a5c561f97b52c17a19f34c4499a745cd4e8412e29e4ed5e91e4481c7d53935"`;
